POWER9 architecture support : Add initial support to create a powerpc64le boostrap

adevress opened the pull request:
- ppc64le is the architecture used for POWER8 / POWER9 machines
- ppc64le implies a bit of trickery for long long double being 128bits

###### Motivation for this change

- Initial request to add POWER9 architecture support

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Ericson2314 commented on lib/systems/inspect.nix:
```diff
@@ -11,6 +11,7 @@ rec {
     isi686         = { cpu = cpuTypes.i686; };
     isx86_64       = { cpu = cpuTypes.x86_64; };
     isPowerPC      = { cpu = cpuTypes.powerpc; };
+    isppc64le      = { cpu = cpuTypes.powerpc64le; };
```
Hmm since this shorthand exists just at the nix level (in this PR) at least, can you capitalize it?

Ericson2314 commented on pkgs/development/compilers/gcc/7/default.nix:
```diff
@@ -88,7 +88,10 @@ let version = "7.3.0";
-      ] else [
+      ] ++ stdenv.lib.optionals ( targetPlatform.config == "powerpc64le-unknown-linux-gnu" )  [
+	"--with-long-double-128"
+	"--enable-softfloat"
```
See what I did with the ABIs in the nixpkgs lib to lessen the cominatorial explosion in GCC. Can you use the existing functionality for soft floats, and make a new one for longer floats?

Ericson2314 commented on pkgs/os-specific/linux/musl/default.nix:
```diff
@@ -59,6 +59,10 @@ stdenv.mkDerivation rec {
   ];
   preConfigure = ''
     configureFlagsArray+=("--syslibdir=$out/lib")
+
```
Please make this conditional on the nix level to avoid rebuilds and be easier to read. Also use `+=` in bash.

Ericson2314 commented on pkgs/stdenv/generic/default.nix:
```diff
@@ -122,7 +122,7 @@ let
       inherit (hostPlatform)
         isDarwin isLinux isSunOS isHurd isCygwin isFreeBSD isOpenBSD
-        isi686 isx86_64 is64bit isAarch32 isAarch64 isMips isBigEndian;
+        isi686 isx86_64 is64bit isArm isAarch64 isMips isBigEndian isppc64le;
```
Looks like an `isArm` snuck back in there? Please switch back to `isAarch32`.

Ericson2314 reviewed (changes requested):
A few things to do, but very excited people are adding more platforms!

adevress committed:
Add initial support to create a powerpc64le boostrap

- ppc64le is the architecture used for POWER8 / POWER9 machines
- ppc64le implies a bit of trickery for long long double being 128bits

adevress commented on lib/systems/parse.nix:
```diff
@@ -89,7 +89,7 @@ rec {
 
     powerpc  = { bits = 32; significantByte = bigEndian;    family = "power"; };
-
+    powerpc64le  = { bits = 64; significantByte = littleEndian;    family = "power"; };
```
No, powerpc64 is a different architecture ( POWER <= 7 ) that I do not own any hardware for. powerle does not exist I think.
powerpc64le is the usual code name used by distribution for OpenPOWER power8/9 generation. meaning PowerPC architecture, 64 bits, little endian

Ericson2314 replied:
hehe the http://llvm.org/doxygen/Triple_8cpp_source.html is inconsistent (see the `get*Variant` functions).

adevress commented on lib/systems/examples.nix:
```diff
@@ -29,6 +29,13 @@ rec {
 
+  powerpc64le-multiplatform = rec {
+    config = "powerpc64le-unknown-linux-gnu";
+    arch = "ppc64le";
```
I started to do this support on a nixpkgs of some month ago... let me remove that

Ericson2314 commented:
#45340 is now merged. We should see what is left here that isn't included in that.
